zcash_primitives/src/transaction/builder.rs

```diff
@@ -625,55 +961,36 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
     /// Builds a transaction from the configured spends and outputs.
     ///
     /// Upon success, returns a tuple containing the final transaction, and the
-    /// [`TransactionMetadata`] generated during the build process.
+    /// [`SaplingMetadata`] generated during the build process.
     ///
     /// `consensus_branch_id` must be valid for the block height that this transaction is
     /// targeting. An invalid `consensus_branch_id` will *not* result in an error from
     /// this function, and instead will generate a transaction that will be rejected by
     /// the network.
     pub fn build(
         mut self,
+        version: TxVersion,
         consensus_branch_id: consensus::BranchId,
         prover: &impl TxProver,
-    ) -> Result<(Transaction, TransactionMetadata), Error> {
-        let mut tx_metadata = TransactionMetadata::new();
-
+    ) -> Result<(Transaction, SaplingMetadata), Error> {
         //
         // Consistency checks
         //
 
         // Valid change
-        let change = self.mtx.value_balance - self.fee
-            + self
-                .transparent_inputs
-                .value_sum()
-                .ok_or(Error::InvalidAmount)?
-            - self
-                .mtx
-                .vout
-                .iter()
-                .map(|vo| vo.value)
-                .sum::<Option<Amount>>()
-                .ok_or(Error::InvalidAmount)?;
+        let change = self
+            .transparent_builder
+            .value_balance()
+            .and_then(|ta| ta + self.sapling_builder.value_balance())
+            .and_then(|b| b - self.fee)
+            .ok_or(Error::InvalidAmount)?;
 
         #[cfg(feature = "zfuture")]
-        let change = change
-            + self
-                .tze_inputs
-                .builders
-                .iter()
-                .map(|ein| ein.prevout.value)
-                .sum::<Option<Amount>>()
-                .ok_or(Error::InvalidAmount)?
-            - self
-                .mtx
-                .tze_outputs
-                .iter()
-                .map(|tzo| tzo.value)
-                .sum::<Option<Amount>>()
-                .ok_or(Error::InvalidAmount)?;
-
-        let change = change.ok_or(Error::InvalidAmount)?;
+        let change = self
```

Comment:
This is overwriting `change`:
```suggestion
        let change = change
          + self
```